documentation/correctness.adoc

```diff
@@ -0,0 +1,106 @@
+= Model correctness
+
+Whether a model is _correct_ –- or _valid_, or _consistent_ –- is an important question.
+Correctness ultimately determines whether a model is useful in the context it exists in.
+Answering that question only with a simple "`yes`" or "`no`" is too simplistic: a model can be incorrect in different ways and on different levels.
+
+Not being able to resolve a reference is a different kind of failure than violating a semantic constraint.
+Having an illegal extra comma in an array somewhere in the JSON serialization chunk of a model -- e.g. `[ 0, 1, 2**,** ]` -- should not prevent that serialization chunk from being deserializable.
+
+[IMPORTANT]
+====
+We'll talk about _correctness_ from now, and consider _validity_ and _consistency_ to be equivalent concepts.
+====
+
+It makes sense to phrase the answer as a collection of _issues_, instead of giving a hard "`yes/no`" answer.
+An issue is a specific point in the model where a specific kind of incorrectness arises.
+Many (kinds of) issues can be meaningfully recovered from, and recovery can often be obvious to the point it can be automatic.
+The unresolved reference can be left in as-is, and treated as an unknown (but not a _missing_) value with special meaning (i.e., domain-specific semantics).
+The illegal comma should be reported but can subsequently be safely ignored.
+In cases where recovery is not automatic, an issue should be very clear to the user what they can do to resolve it.
+A model must be editable/mutable -- and even usable to some -- even if it has issues.
+
+LionWeb provides a precise notion for model correctness in the form of a multi-leveled specification.
+For each level, it specifies what kinds of incorrectness can arise, how these should be reported as issues, and whether and how these can be meaningfully _recovered_ from.
+This approach provides a degree of resilience: issues arising in one level don't necessarily prohibit determining whether a model exhibits incorrectness on another level, provided an issue can be meaningfully _recovered_ from.
```

Comment:
I am not sure about the term _recovered_. Some issues can be completely recovered from (e.g., missing commas or extra commas in JSON), in the sense that we can operate as if the issue was not there in the first place. Other issues can not be completely _recovered_ from. For example, references that cannot be recovered and are marked as having value unknown are not fully _recovered_ as we cannot get the exact behavior as we would if the reference was solvable. We get some sort-of-working state, that requires extra care from the user (who now needs to handle the unknown values) and some operations cannot be performed at all, because the issue can only be partially recovered.

I would consider using a term different from recovered to indicate all the issues that do not cause a critical stop in loading the model (i.e., if get an XML file instead of a JSON file there is no way to provide even a partial model). Then we could differentiate between issues that can be fully recovered (like extra commas), and issues for which we have a workaround (like using the unknown value)

Reply:
Would fully vs. partially recoverable (instead of ignorable vs. stubbable – see other comment) work for you?

Reply:
Yes, thank you